x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java

```diff
@@ -0,0 +1,131 @@
+        }
+        final BigArrays bigArrays = context.bigArrays();
+        MultiGeoPointValues docGeoPointValues = valuesSources.getGeoPointField(GEO_POINT_FIELD.getPreferredName(), ctx);
+        SortedNumericDoubleValues docSortValues = valuesSources.getNumericField(SORT_FIELD.getPreferredName(), ctx);
+
+        return new LeafBucketCollectorBase(sub, docGeoPointValues) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                positions = bigArrays.grow(positions, bucket + 1);
+                if (docGeoPointValues.advanceExact(doc) && docSortValues.advanceExact(doc)) {
+                    if (docSortValues.docValueCount() > 1) {
+                        throw new AggregationExecutionException("Encountered more than one sort value for a " +
+                            "single document. Use a script to combine multiple sort-values-per-doc into a single value.");
+                    }
+                    if (docGeoPointValues.docValueCount() > 1) {
+                        throw new AggregationExecutionException("Encountered more than one geo_point value for a " +
+                            "single document. Use a script to combine multiple geo_point-values-per-doc into a single value.");
+                    }
+
+                    // There should always be one weight if advanceExact lands us here, either
+                    // a real weight or a `missing` weight
+                    assert docSortValues.docValueCount() == 1;
+                    assert docGeoPointValues.docValueCount() == 1;
+                    final double sort = docSortValues.nextValue();
+                    final GeoPoint point = docGeoPointValues.nextValue();
+                    ArrayList<InternalGeoLine.Position> bucketLine = positions.get(bucket);
+                    if (bucketLine == null) {
+                        bucketLine = new ArrayList<>(1000);
+                    } else if (bucketLine.size() >= 10000) {
+                        throw new AggregationExecutionException("Cannot aggregate a line with more than 10000 points");
+                    }
+                    bucketLine.add(new InternalGeoLine.Position(
+                        GeoEncodingUtils.encodeLatitude(point.lat()),
+                        GeoEncodingUtils.encodeLongitude(point.lon()), sort));
+                    positions.set(bucket, bucketLine);
+                }
+            }
+        };
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) {
+        if (valuesSources == null) {
+            return buildEmptyAggregation();
+        }
+        List<InternalGeoLine.Position> bucketLine = positions.get(bucket);
+        if (bucketLine == null) {
+            bucketLine = Collections.emptyList();
+        } else {
+            Collections.sort(bucketLine);
+        }
+        return new InternalGeoLine(name, bucketLine, pipelineAggregators(), metaData());
```

Comment:
This is probably over-optimizing but thought I'd mention:

Since the points are sorted by time, the x/y deltas should be reasonably small between consecutive points (in the general case).  We could, theoretically, re-compress the line using delta-encoding and it might save a bunch of space.  It would probably complicate the reduction on coordinating node though.